File: frag.py
import sys
import subprocess
import os

dummyFile_f = os.open("/mnt/dummy", os.O_CREAT|os.O_DIRECT|os.O_RDWR)
targetFile_f = os.open(sys.argv[1], os.O_DIRECT|os.O_RDWR)

fragUnit=4096
times=1

while len(data) > 0:
    mmap.mmap(dummyFile_f, fragUnit,-1, 1024 * fragUnit)
    
    os.read(

    m.write(data)
    os.seek(dummyFile_f, -1 * fragUnit)
    os.write(dummyFile_f, m)

    data = so.read(targetFile_F, fragUnit)
    os.write(targetFile_f, fragUnit)

frag_degree.seek(0)
frag_degree.readline()
line = frag_degree.readline()
filesize = line.split(' ')[5]
frag_degree.readline()


lines = frag_degree.readlines()


bufsize = 0
need = 0
for line in lines[:-1]:
    fragsize = int(line.split(':')[3])
    bufsize += fragsize * 4
    if bufsize < defragsize:
        need = 1
        continue

    while bufsize >= defragsize:
        if need == 1:
            '''print(need)'''
            data = target_file.read(defragsize*1024)
            size = len(data)
            target_file.seek(-1*size,1)
            target_file.write(data)

            '''os.fsync(target_file.fileno())'''
            need = 0
            target_file.seek(-1*defragsize*1024,1)

        
        target_file.seek(defragsize*1024,1)
        bufsize -= defragsize

    if bufsize > 0:
        need=1

'''print(target_file.tell())'''
if need == 1:
    data = target_file.read(bufsize*1024)
    size = len(data)
    target_file.seek(-1*size,1) 
    target_file.write(data)
    
target_file.flush()
os.fsync(target_file.fileno())

target_file.close()
frag_degree.close()

File: frag.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/file.h>
#include <sys/types.h>
#include <sys/stat.h> 
#include <string.h>

#include <fcntl.h>
#define BUF_SIZE 4096
//#define _GNU_SOURCE

int main()
{
	int fd, fd2, fd3;
	int ret;
	unsigned char *buf;
	ret = posix_memalign((void **)&buf, 4096, BUF_SIZE);
	if (ret) {
		perror("posix_memalign failed");
		exit(1);
	}
	fd = open("/mnt/a.c", O_RDWR | O_DIRECT | O_SYNC | O_CREAT, 0755);
	fd2 = open("/mnt/tmp", O_RDWR | O_DIRECT | O_SYNC | O_CREAT, 0755);
	fd3 = open("/mnt/dummy", O_RDWR | O_DIRECT | O_SYNC | O_CREAT, 0755);
	if (fd < 0){
		perror("open failed");
		exit(1);
	}

	do {
		ret = read(fd, buf, BUF_SIZE);
		if (ret == 0)
			break;
		if (ret < 0) 
			perror("read ./direct_io.data failed");
		ret = write(fd2, buf, ret);
		write(fd3, buf, ret);
		printf("%d\n",ret);
		if (ret < 0) 
			perror("write ./direct_io.data failed");
	} while (ret > 0);
	
	printf("finished \n");
	close(fd);
	close(fd2);

	rename("/mnt/tmp", "/mnt/a.c");

	free(buf);
}


